### ISSUE TITLE
Network policy ingress rule can't match remote non-k8s-node IP

### ISSUE THREAD

**Issue by tungdam** [projectcalico/calico#2660]
Hi
I have a problem with applying network policy for my pod against the node which is outside the k8s cluster. 

The situation is like this:

We have kubernetes 1.12, calico 3.5 setup, which advertise clusterIP range over BGP following [this](https://docs.projectcalico.org/v3.5/usage/service-advertisement) instruction. 

As security concern, we just want to expose my pod to fixed list of remote hosts ( non-k8s ones ). 
Here's my network policy config: 

```
apiVersion: networking.k8s.io/v1
kind: NetworkPolicy
metadata:
  name: test-policy
  namespace: default
spec:
  podSelector:
      matchLabels:
          svc-group: test
  policyTypes:
  - Ingress
  ingress:
  - from:
    - ipBlock:
        cidr: 10.101.1.1/32
    ports:
    - protocol: TCP
      port: 80
```

My svc/deployment yaml: 
```
apiVersion: apps/v1
kind: Deployment
metadata:
  labels:
    app.k8s.itim.vn/name: "td-nginx"
  name: "td-nginx"
  namespace: default
spec:
  replicas: 2
  selector:
    matchLabels:
      application: "td-nginx"
  template:
    metadata:
      annotations:
        cni.projectcalico.org/ipv4pools: '["10.199.0.0/21"]'
      labels:
        application: "td-nginx"
        svc-group: "test"
    spec:
      containers:
      - name: nginx
        image: nginx
        ports:
        - containerPort: 80
          protocol: TCP
---
apiVersion: v1
kind: Service
metadata:
  labels:
    app.k8s.itim.vn/name: "td-nginx"
    app.k8s.itim.vn/instance: "td-nginx"
  name: "td-nginx"
  namespace: default
spec:
  type: ClusterIP
  ports:
  - port: 80
    targetPort: 80
  selector:
    application: "td-nginx"
```

My test service's clusterIP:
```
kubectl get svc td-nginx
NAME       TYPE        CLUSTER-IP     EXTERNAL-IP   PORT(S)   AGE
td-nginx   ClusterIP   10.196.54.69   <none>        80/TCP    27h
```

My clusterIP range: 10.196.48.0/20 
My calico pod ip range: 10.198.0.0/15 ( which i set to kube proxy's clusterCIDR param ) 
IPs for test pod 
```
kubectl get pod -o wide | fgrep td-nginx
td-nginx-6658b757c-5lxsw                             1/1     Running   0          27h   10.199.7.64    nodeA          <none>
td-nginx-6658b757c-c5snz                             1/1     Running   0          27h   10.199.5.129   nodeB               <none>
```
## Expected Behavior

With network policy above applied: 

Can access to port 80 of my service by clusterIP ( 10.196.54.69 ) from an IP ( non-k8s node which is white-listed in network policy config - 10.101.1.1. 

## Current Behavior
With network policy applied: 
- Can't access from external node ( 10.101.1.1 ) to clusterIP ( 10.196.54.69 ) 
- Still can access to pod IP directly 
- if we reconfigure service to LoadBalancer type as below
```
spec:
  type: LoadBalancer
  externalTrafficPolicy: Local
```
And edit deployment to deploy 1 replica only, it works. 
If replica = 2, sometimes it works, sometimes it's blocked 

With network policy removed:
- Can access to both clusterIP and pod IPs

## Possible reason 
The source IP before hitting network policy was changed by kube-proxy
`-A KUBE-SERVICES ! -s 10.198.0.0/15 -d 10.196.54.69/32 -p tcp -m comment --comment "default/td-nginx: cluster IP" -m tcp --dport 80 -j KUBE-MARK-MASQ`

## Question
Is it possible to apply network policy with external IP address ( outside kubernetes cluster)  as I did above ? 

Thanks in advance 

## My Environment
* Calico 3.5
* Kubernetes 1.12
* Debian 9

**Comment by tmjd** [comment]
Can you check on the routes that your host is seeing for the service IP? Is it receiving a route for a host that is not running the destination pod?
If it is only receiving routes to hosts that have a backing pod for the service IP, then you will need to look at kube-proxy and service configuration to figure out why it is deciding to SNAT the traffic (when it does).

**Comment by tungdam** [comment]
> Can you check on the routes that your host is seeing for the service IP? Is it receiving a route for a host that is not running the destination pod?
As i posted in the description, we config our svc like this:
```
spec:
  type: LoadBalancer
  externalTrafficPolicy: Local
```
So it always goes to the host that have a backing pod. But when we have multiple replicas, kube-proxy will send the request randomly like this:

```
-A KUBE-SVC-23S72TPQW2AY74ZS -m statistic --mode random --probability 0.50000000000 -j KUBE-SEP-APFTNSXX6TWBJEXE
-A KUBE-SVC-23S72TPQW2AY74ZS -j KUBE-SEP-JN6YB2LIAAJICJGV
```
Sometimes our request goes to local pod -> it works as it matches rule for local cali+ interface
Sometimes it is sent to a remote pod on another host -> doesn't work. 

> If it is only receiving routes to hosts that have a backing pod for the service IP, then you will need to look at kube-proxy and service configuration to figure out why it is deciding to SNAT the traffic (when it does).

Because kube-proxy always SNAT with the IP range that is not pod IP range ( which we set by clusterCIDR param ), the request from the external host will be SNAT-ed, IMHO. 

A workaround for this case is somehow we can set multiple subnets/ IP range for clusterCIDR param so both pod IP range and our physical hosts IP range won't be SNAT-ed. But AFAIK, it's not possible. 

Hope that i got your question correctly and answered it well.

**Comment by tmjd** [comment]
I think you might need to switch to `type: NodePort`.
If I'm reading it correctly https://kubernetes.io/docs/tutorials/services/source-ip/#source-ip-for-services-with-type-loadbalancer indicates in the section "Cross platform support" that source IP preservation is only available on GCP and Azure

**Comment by rafaelvanoni** [comment]
Closing due to inactivity, please provide an update if this is still an issue and we'll re-open it.

**Comment by tungdam** [comment]
Sorry for taking it too long to update. We still need it. We don't want to use NodePort because we prefer having nice standard 80 or 443 ports for our services ( everyone does, i guess ). 
Beside the limitation with network policy as explained above , the problem's getting bigger with us when we have some services that requires the correct source IP of client for its ACL. 

So far we have to use headless services ( so clients connect directly to Pod IP ) but it's not a complete solution as well. 

( Be cause service discovery depends heavily on dns round robin. Some clients have their own dns caching mechanism that may end up resolving to an obsolete pod IP ) 

I really love the  service cluster IP advertising feature by calico, it helps us to expose services on our on-prem infrastructure to the outside very well but the source IP problem now is very big drawback to us. 

Please consider supporting source IP preserving somehow.

**Comment by tmjd** [comment]
Have you tried with a NodePort? I believe with node ports you still get a service IP and can specify the port that is used with the Service IP (you can still use 443 or 80). This [page](https://kubernetes.io/docs/concepts/services-networking/connect-applications-service/#exposing-the-service) seems to back up that idea too.

**Comment by tungdam** [comment]
@tmjd Thanks for your quick reply 

I tried, but NodePort can't solve my problem because:
- There will be port conflicting issue when most likely we'll need more than 1 service requires binding to 80/443 
- NodePort still do source NAT by default. It only preserves source IP if we set `"externalTrafficPolicy":"Local"` but then we have to remember by ourselves on which node the pod is deployed to ( and what's its NodePort ). It sacrifices all the service discovery convenience. 

The key thing we would like to have here is we still can benefit from clusterIP ( and the coreDNS record to load balance, service discovery ) AND preserve the source IP at the same time. 

Is it possible ?

**Comment by tmjd** [comment]
It sounds like you are assuming that when you use NodePort that the clusterIP of a service is no longer usable, as far as I know that is not true. When you use a NodePort you still get the clusterIP and will have the coreDNS record and service discovery (you may lose some load balancing but you do not want your connections to be redirected/SNATed so I think that is necessary).

> There will be port conflicting issue when most likely we'll need more than 1 service requires binding to 80/443

Each service will get its own clusterIP so you should be able to have any number of services using 80/443 since they will all be on different IPs. I wasn't suggesting setting the NodePort (nodePort field) but setting the port field ( for reference https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.16/#serviceport-v1-core), in your case you would most likely not set NodePort and let Kubernetes select one.

>NodePort still do source NAT by default. It only preserves source IP if we set "externalTrafficPolicy":"Local" but then we have to remember by ourselves on which node the pod is deployed to ( and what's its NodePort ). It sacrifices all the service discovery convenience.

If you are using the clusterIP for access then you do not need to remember which nodes the pod is deployed to or remember its NodePort.

Have you tried accessing a service of type NodePort through its clusterIP?

@caseydavenport am I wrong here in my understanding of NodePort services?

**Comment by tmjd** [comment]
@tungdam I was just chatting in the Calico Community meeting and found out that a LoadBalancer is essentially an additional layer on a NodePort and should also work for exposing a service and allow the destination pods to receive the Source IP of traffic.
@neiljerram would you have any suggestions here? It sounded like there is perhaps another issue where you're doing similar investigation.

**Comment by tungdam** [comment]
I understand that NodePort type service still have clusterIP and coreDNS but such clusterIP is no different than the normal clusterIP in our case. Source IP is not preserved

**Comment by tungdam** [comment]
> @tungdam I was just chatting in the Calico Community meeting and found out that a LoadBalancer is essentially an additional layer on a NodePort and should also work for exposing a service and allow the destination pods to receive the Source IP of traffic.
Yes Loadbalancer type can preserve srcIP but as i explained in the description of this issue, if we have >=2 pods which is deployed to =>2 different nodes, it's not true anymore.

**Comment by caseydavenport** [comment]
This really sounds like an issue with kube-proxy, who is responsible for preserving the source address when handling externalTrafficPolicy: Local services.

**Comment by sourabhgupta385** [comment]
What is the proposed solution to this? 
My use-case - 

We have redis deployed in one kubernetes cluster in region A and to implement DR scenario, we have redis deployed in another cluster in another region B. Now, to implement the sync between two redis pods, we want to allow ingress in A region redis pods from B region redis pods. 

I wrote a network policy which allowed ingress from CIDR of worker nodes of B region. That did not worked out. When I enabled logging in region A, I found that all the external requests are reaching to redis pods with SRC IP as 10.244.9.0 which kinda feels like gateway IP. 

How are we supposed to preserve the original IPs only?
